rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,399 @@
+processed.
+
+With indexer acknowledgement enabled, the `200 OK` response contains an
+additional JSON body field specifying an `ackId` where `ackId` is an integer
+identifier corresponding to the request. Note that a single HEC request can
+include multiple Splunk events, so the `ackId` covers the entire request.
+
+```jsonc
+// Example response from local Splunk 7.3.2
+{
+   "text": "Success",
+   "code": 0,
+   "ackId": 10
+}
+```
+
+Using one or more `ackId`’s, the user can query Splunk’s
+`/services/collector/ack` endpoint to check on the status of associated
+requests. For each `ackId`, Splunk returns `true`/`false` depending on whether
+the data in the request has been persisted (in Splunk’s words, “replicated at
+the desired replication factor”). Upon returning `true`, Splunk drops the
+`ackId` status and all subsequent requests with the same `ackId` will return
+`false` (based on Splunk channel-related expiration settings, `ackId`'s can be
+reset/reused).
+
+```jsonc
+// Example request body
+{
+    "acks": [0, 1, 2]
+}
+
+// Example response body
+{
+    "acks":
+        {
+            "0": true,
+            "1": false,
+            "2": true
+        }
+}
+```
+
+In addition to the overall protocol, there are a few details worth highlighting:
+
+* Authentication tokens
+  * To submit requests to HEC in general, an authentication token must be
+    included in request headers. Indexer acknowledgement is enabled per
+    authentication token in Splunk. The `splunk_hec` source currently supports a
+    list of valid authentication tokens. The `splunk_hec` sinks currently
+    support configuring a single authentication token.
+
+* Splunk channels
+  * Upon enabling indexer acknowledgement, Splunk requires HEC requests to
+    include a channel ID (any valid `guid` value). Querying
+    `/services/collector/ack` also requires a channel ID.
+  * `ackId`’s are assigned per-channel.
+
+* Pending acks
+  * Clients are not required or guaranteed to query for `ackId` status. To avoid
+    running out of memory from `ackId` build up, Splunk offers a few
+    configuration options to limit the number of pending `ackId`’s both overall
+    and per-channel.
+  * Splunk does not explicitly indicate whether event data is dropped or still
+    pending processing. Either of these states will result in `{ack_id}: false`
+    status. Splunk only advises that after a certain amount of time (e.g. 5
+    minutes), the data can be considered dropped.
+
+## Proposal: `splunk_hec` Source Indexer Acknowledgement
+
+### User Experience
+
+Users can configure the `splunk_hec` source with additional indexer
+acknowledgement related settings.
+
+```toml
+[sources.splunk_hec]
+type = "splunk_hec"
+# ...
+acknowledgements.enabled = true
+acknowledgements.max_pending_acks = 10_000_000
+acknowledgements.max_number_of_ack_channel = 1_000_000
+acknowledgements.max_pending_acks_per_channel = 1_000_000
```

Comment:
👍 for matching Splunk here.

Reply:
On that note, would we prefer to keep Splunk's naming convention? I did shorten some option names (e.g. `max_number_of_acked_requests_pending_query_per_ack_channel` to `max_pending_acks`).

Reply:
I'd probably name it the same since what we are looking to do here is mostly let users familiar with Splunk migrate over. At the least, we can reference the Splunk name in the docs if we want to shorten it.

Reply:
My only objection here is just how long the name is :/

Reply:
I'm game for referencing the Splunk name in docs and keeping the shortened version.